README
README for Image::Compare

			    Image::Compare

Image::Compare is a module for performaing comparisons of, well..  images.

PREREQUISITES

This suite requires Perl 5; I've only used it under Perl 5.8.8, so for
anything lower, you're on your own.

Image::Compare makes heavy use of Imager, and also can make use of LWP if
it is installed.

INSTALLATION

You install Image::Compare, as you would install any perl module
library, by running these commands:

   perl Makefile.PL
   make
   make test
   make install

If you want to install a private copy of Tree::Trie in your home
directory, then you should try to produce the initial Makefile with
something like this command:

  perl Makefile.PL PREFIX=~/perl

DOCUMENTATION

POD-format documentation is included in Compare.pm.  POD is readable
with the 'perldoc' utility.

SUPPORT

Questions, bug reports, useful code bits, and suggestions for
Image::Compare should just be sent to me at avi@finkel.org

AVAILABILITY

The latest version of Image::Compare is available from the
Comprehensive Perl Archive Network (CPAN).  Visit
<http://www.perl.com/CPAN/> to find a CPAN site near you.

COPYRIGHT

Copyright 2008, Avi Finkel <avi@finkel.org>, all rights
reserved.  This program is free software; you can redistribute it and/or 
modify it under the same terms as Perl itself.


Changes
Change file for Image::Compare
Avi Finkel <avi@finkel.org>

Version 0.1 : 4/11/06 15:56

   No changes -- initial release

Version 0.2 : 7/23/06 01:45

   Adding a few tests

Version 0.3 : 9/15/06 01:45

   Adding accessors for image1, image2 and method
	 Fixing up makefile to make it friendlier

Version 0.5 : 4/6/08 14:45

   Major refactoring
	 Made some requirements into recommendations
	 Added new THRESHOLD_COUNT comparator
	 Added support for user-provided color gradients
   Added match mask support
	 I'm not sure where 0.4 went.

Version 0.6 : 5/1/10 00:32

	Fixing some Pod::Coverage problems

Version 0.7 : 5/18/10 20:34

	Making the dumb network test marginally less moronic

Version 0.8 : 5/29/10 12:22

	Yet ANOTHER fix for the world's worst test.

Version 0.9 : 9/23/10 22:33

	I really hope this is the last time I have to fix this test.

Version 1.0: 12/16/14 1:06

    Adding some clarity to the documentation around the IMAGE output method.
    Thanks to Daniel Gempesaw for the patch!


t/helper.pm
# Helper methods for making images etc

use warnings;
use strict;

sub make_image {
	my $img = Imager->new(
		xsize => scalar(@_),
		ysize => scalar(@{$_[0]}),
	);
	for my $x (0 .. $#_) {
		for my $y (0 .. $#{$_[0]}) {
			$img->setpixel(x => $x, y => $y, color => [($_[$x][$y]) x 3]);
		}
	}
	return $img;
}

1;


lib/Image/Compare/Comparator.pm
package Image::Compare::Comparator;

use warnings;
use strict;

sub new {
	my $proto = shift;
	my $class = ref($proto) || $proto;
	my $self = {};
	$self->{args} = shift;
	$self->{mask} = shift;
	bless($self, $class);
	return $self;
}

# This will do initial setup and throw an exception if there is something
# wrong.  We have some common behavior in here.  Subclasses may override this,
# or add to it.
sub setup {
	my $self = shift;
	my ($img1, $img2, $mask) = @_;
	unless (
		($img1->getwidth()  == $img2->getwidth() ) &&
		($img1->getheight() == $img2->getheight())
	) {
		die "Images must be the same size!";
	}
	if ($mask) {
		unless(ref($mask) && $mask->isa('Imager')) {
			die "Match mask must be an Imager image object!";
		}
		unless (
			($mask->getchannels() == 1) &&
			($mask->bits()     == 8)
		) {
			die "Match mask image must have one channel and 8 bits per channel!";
		}
		unless (
			($mask->getwidth()  == $img1->getwidth() ) &&
			($mask->getheight() == $img1->getheight())
		) {
			die "Match mask must be the same size as the test images!";
		}
	}
}

sub get_args {
	my $self = shift;
	return $self->{args};
}

# By default, just return the class name and the arguments.
sub get_representation {
	my $self = shift;
	return (
		method => $Image::Compare::reverse_class_map{ref($self)},
		args => $self->{args},
	);
}

sub compare_images {
	my $self = shift;
	my ($img1, $img2, $mask) = @_;
	# This will die if there's a problem.
	$self->setup($img1, $img2, $mask);
	# We spin over each pixel in img1.
	my $wid = $img1->getwidth();
	my $hig = $img1->getheight();
	OUTER: for my $x (0 .. $wid - 1) {
		for my $y (0 .. $hig - 1) {
			# If we've been given a match mask, then we skip any pixel whose
			# corresponding pixel in that mask is pure black.
			# This is the entirety of the comparison logic surrounding masks.  It is
			# all so simple, I should have done it long ago.
			if ($mask && (($mask->getpixel(x => $x, y => $y)->rgba())[0] == 255)) {
				next;
			}
			my @pix1 = $img1->getpixel(x => $x, y => $y)->rgba();
			my @pix2 = $self->get_second_pixel($img2, $x, $y)->rgba();
			# If this returns undef, then we keep going.  Otherwise, we stop.
			# It will die if there's an error.
			# This mechanism allows the subclass to short-circuit image examination
			# if it feels the need to do so.
			last OUTER if defined $self->accumulate(\@pix1, \@pix2, $x, $y);
		}
	}
	# And finally, the subclass will return the thing it wants to return.
	return $self->get_result();
}

# By default, this is pretty boring.
# Subclasses may want to override it though.
# On second thought, I can't think of a reason why they would want to.
# I guess I will leave this in anyways.
sub get_second_pixel {
	my $self = shift;
	my ($img2, $x, $y) = @_;
	return $img2->getpixel(x => $x, y => $y);
}

# Some day we might have multiple ways to do this.
sub color_distance {
	my $self = shift;
	my ($pix1, $pix2) = @_;
	# The sum of the squaws of the other two hides...
	return sqrt(
		( ($pix1->[0] - $pix2->[0]) ** 2 ) +
		( ($pix1->[1] - $pix2->[1]) ** 2 ) +
		( ($pix1->[2] - $pix2->[2]) ** 2 )
	);
}

sub accumulate {
	my $self = shift;
	my $class = ref($self) || $self;
	die "Subclass '$class' must implement accumulate()!";
}

sub get_result {
	my $self = shift;
	my $class = ref($self) || $self;
	die "Subclass '$class' must implement get_result()!";
}

sub import {
	my $cmp_pkg = shift;
	my %args = @_;
	unless (UNIVERSAL::isa($cmp_pkg, __PACKAGE__)) {
		die "Comparaters must subclass __PACKAGE__!";
	}
	my $name = $cmp_pkg;
	unless (
		($name =~ s/^Image::Compare:://) ||
		($name = $args{name})
	) {
		die (
			"Comparator must either be in the Image::Compare namespace, " .
			"or you must provide a method name to import."
		);
	}
	{
		no strict qw/refs/;
		# We are essentially "exporting" this for backwards compatibility.  We
		# don't really want to use constants like this any more, but we have
		# to.  Shucks.
		*{"Image::Compare::$name"} = sub () { $name };
		$Image::Compare::class_map{$name} = $cmp_pkg;
		$Image::Compare::reverse_class_map{$cmp_pkg} = $name;
	}
}

# We will read in the list of packages to load from the documentation.
while (<Image::Compare::Comparator::DATA>) {
	if (/^=item \* L<([^>]+)>/) {
		eval "use $1";
		die "Failed loading module '$1': $@" if $@;
	}
}

close Image::Compare::Comparator::DATA;

1;

__DATA__

=pod

=head1 NAME

Image::Compare::Comparator - Base class for comparison methods used by Image::Compare

=head1 OVERVIEW

This is essentially an abstract class which defines some basic functionality
and outlines some patterns for use of subclasses which will each define a
different process for comparing two images. The documentation here is aimed
more for those wishing to write their own comparators.

=head1 COMPARATORS

See each submodule's documentation for information about how it works.

=over 4

=item * L<Image::Compare::AVG_THRESHOLD>

=item * L<Image::Compare::EXACT>

=item * L<Image::Compare::THRESHOLD>

=item * L<Image::Compare::THRESHOLD_COUNT>

=item * L<Image::Compare::IMAGE>

=back

=head1 CREATING CUSTOM COMPARATORS

=over 4

=item OVERVIEW

There are 5 methods that can be overriden by comparator subclasses, and two
that must be.  Most of the work in creating your owm comparator is in the
implementation of your comparison logic, which is out of the purview of
this document, so it will be pretty short.

=item IMPORT

It is optional to override the import() method.  If you do, you must call the
superclass import().  If your comparator is not in the Image::Compare::
namespace, then you must be sure that the superclass import() is called with
an argument called "name", containing a unique name by which the comparator
will be referred in the system.  If the comparator B<is> in the
Image::Compare:: namespace, then the name will be assumed to be the class
name, but that can be overriden using the name argument to import().  For
example:

 package My::Comparator::FOO

 use base qw/Image::Compare::Comparator/;

 sub import {
	 my $class = shift;
	 $class->SUPER::import(name => 'FOO');
	 # ... Whatever
 }

You should override import if:

=over 4

=item *

You need to do some special logic at class initialization time.

=item *

Your class isn't in the Image::Compare:: namespace and you need to provide the
comparator name.

=item *

You want to provide a different comparator name, even though your comparator
is in the Image::Compare:: namespace.

=back

=item SETUP

It is optional to override the setup() method.  If left alone, it will do some
very basic input checking.  If you have any logic which must run before each
comparison run begins, this is where you'd put it.  You should put any pre-run
sanity checking code here, and die() if there's a problem.  You need not call
the superclass version of this method if you don't want to, but you may in
order to leverage the error checking logic there.

=item ACCUMULATE

This method must be overriden and it will contain most of your subclass's
logic.  Its inputs will be:

=over 4

=item *

A reference to an array containing the red, green and blue values of a certain
pixel in the first image.

=item *

A reference to an array containing the red, green and blue values of a certain
pixel in the second image.

=item *

The X coordinate whence came the color values from the first image.  In most
cases, this will be the same as the X coordinate for the second image, but if
you've overriden get_second_pixel(), then this may not be true.

=item *

The Y coordinate whence came the color values from the first image.  The same
caveat applies here as for X.

Your logic should do whatever calculations are neccessary with this data and
store whatever information must be stored for further processing.  If you
can determine at any point that further processing is wasteful, you can return
any non-undef value from this method and processing will cease.  In the usual
case, you should simply return undef to allow processing to continue.  die()
from this method if there's a problem.

=back

=item GET_RESULT

This method will be called at the end of the comparison run, and must be
override by the subclass.  This should do any final work that needs to be
done for the comparison, and should then return whatever value should be
returned to the end user.

=item GET_SECOND_PIXEL

Overriding this method is optional.  By default, if the XxY pixel is read
from the first image then the XxY pixel is also read from the second image, and
they are then compared.  By overriding this method, one can cause the
application to compare the XxY pixel from the first image with the X'xY' pixel
from the second image.

Note: This functionality is experimental and is subject to change as its use
case becomes more clear.

=back

=head1 METHODS

=over 4

=item new

Constructs a new instance of a comparator.  Should not be overridden by
subclasses.  All this does is the basic perl OO stuff, and copies the second
argument for later usage by the subclass.

=item setup($img1, $img2)

Subclasses should override this method to do input checking of their arguments
and die if there are any problems.  By default, this method only verifies that
both images are the same dimensions.

=item accumulate(\@pixel1, \@pixel2, $x, $y)

This method will be called once for each pair of pixels in the two images.
The RGB color values for those pixels will be passed in, along with the X and
Y coordinates of those pixels.  The return value should be undef if processing
should continue; this means that the comparator has recorded what it needs to
record and sees no reason for things to stop.  If, however, the comparator has
decided at this point (for whatever reason) that further processing would not
change its answer, it can return any non-undef value, and that value will be
returned as the result of the comparison.  This short-circuiting can be very
helpful in speeding up processing for certain types of comparison.

This method must be overridden by subclasses of this class.

=item color_distance(\@pixel1, \@pixel2)

Returns the magnitude of the linear (pythagorean) distance between two pixels
in color-space.  In the future, this method may be modifiable to use
different color distance methodologies, but for now only the simplest is
available.

=item compare_images($img1, $img2)

The main entrypoint method for consumers of instances of subclasses of this
class.  This handles all the "wrapping" for comparing two images.  Generally
should not be subclassed.

=item get_args()

Returns the arguments that had been passed to this instance of this
comparator at construction.  The same as get_representation()->{args}.
Included for backwards compatibility.

=item get_representation()

Returns a hash-representation of the current comparator instance.  The
hash currently contains two keys: I<method>, which maps to a constant
which could be used to construct more instances of this comparator, and
I<args>, which maps to the arguments as described in the documentation for
get_args().

=item get_result()

If all of the images' pixels are processed without accumulate() returning
a defined value, then this method will be called and its return value will
be used as the result of the comparison.

This method must be overridden by subclasses of this class.

=item get_second_pixel()

Given the second image and the current x, y coordinates on which processing
of the first image is currently at, returns the x and y coordinates at which
to process the second image.  By default, simply returns x and y.  Subclasses
may override this method in order to change this behavior.

=back

=head1 AUTHOR

Copyright 2008 Avi Finkel <F<avi@finkel.org>>

This package is free software and is provided "as is" without express
or implied warranty.  It may be used, redistributed and/or modified
under the terms of the Perl Artistic License (see
http://www.perl.com/perl/misc/Artistic.html)

=cut


lib/Image/Compare.pm
# Image::Compare, a module based on the great Imager, used to determine if
# two images differ greatly from one another.

package Image::Compare;

use strict;
use warnings;

use Imager;

# This is the base class for all comparators, and will also do the work of
# loading all supplied implementations.
use Image::Compare::Comparator;

our %class_map;

my $loaded_lwp;

BEGIN {
	$loaded_lwp = 0;
	eval "require LWP;";
	unless ($@) { $loaded_lwp = 1; }
}

our $VERSION = "1.0";

# If people don't want to deal with OO, we export the main "work" method
# so they can call it in a simpler way.  We'll see below where we handle this.
use base qw/Exporter/;
our @EXPORT_OK = qw/compare/;

##   Public methods begin here

# The constructor method.
# Takes a hash of arguments:  (all are optional)
#   image1 =>
#     Data representing the first image, either as an Imager object, file
#     name or a URL.
#   type1  => Type of image provided.
#   image2 => Like image1.
#   type2  => Like type1.
#   method =>
#     Either the numeric constant representing the comparator, or an
#     instance of a comparator.
#   args  => Arguments to pass to the comparator.
# See the documentation on the relevant option setters for more details
sub new {
	my $proto = shift;
	my %args = @_;
	my $class = ref($proto) || $proto;  # Bite me, Randal.
	my $self = {};
	bless($self, $class);
	# These are default values
	if ($args{image1}) {
		$self->set_image1(
			img  => $args{image1},
			type => $args{type1}
		);
	}
	if ($args{image2}) {
		$self->set_image2(
			img  => $args{image2},
			type => $args{type2}
		);
	}
	if ($args{method}) {
		$self->set_method(
			method => $args{method},
			args => $args{args}
		);
	}
	if ($args{mask}) {
		$self->set_mask(mask => $args{mask});
	}
	return $self;
}

# The next two just use the input to fetch image data and store it as an
# Imager object.  Currently supported image types:
#   Imager object
#   File name
#   URL
sub set_image1 {
	my $self = shift;
	my %args = @_;
	$self->{_IMG1} = _get_image($args{img}, $args{type});
}

sub set_image2 {
	my $self = shift;
	my %args = @_;
	$self->{_IMG2} = _get_image($args{img}, $args{type});
}

# Get back the Imager objects created by the preceding two methods.
sub get_image1 {
	my $self = shift;
	return $self->{_IMG1};
}
sub get_image2 {
	my $self = shift;
	return $self->{_IMG2};
}

# How to set the matching mask parameter for this compaison instance.
sub set_mask {
	my $self = shift;
	my %args = @_;
	$self->{_MASK} = $args{mask};
}
sub get_mask {
	my $self = shift;
	return $self->{_MASK};
}

# Given input as defined above, returns an Imager object representing the
# image.
sub _get_image {
	my($img, $type) = @_;
	unless ($img) {
		die "Missing 'img' parameter";
	}

	# If we've been given an Imager object, we need only store it.
	if (ref($img)) {
		if ($img->isa('Imager')) {
			return $img;
		}
		# If it wasn't an Imager, but it's still some kind of reference, then
		# we have to give up.
		die "Unrecognized input type: '" . ref($img) . "'";
	}
	
	# Otherwse, we need to construct an Imager object, and to do that, we
	# need to build up an arguments hash for the Imager constructor.
	my %args;
	if ($type) {
		# Provide the type argument to image, if it was provided.
		$args{type} = $type;
	}
	# This is the base error message.
	my $errmsg = "Unable to read image data from ";
	# If $img looks like a URL, and if we were able to load LWP, then we might
	# be able to fetch an image via a URL.
	if ($loaded_lwp && ($img =~ /^https?:\/\//)) {
		$errmsg .= "URL '$img'";
		my $ua = LWP::UserAgent->new();
		$ua->agent("Image::Compare/v$VERSION ");
		my $res = $ua->request(HTTP::Request->new(GET => $img));
		$args{data} = $res->content();
		if (!$type) {
			$args{type} = $res->content_type();
			$args{type} =~ s!^image/!!;
		}
	}
	else {
		# Otherwise, we have to think it's a file path.
		$errmsg .= "file '$img'";
		$args{file} = $img;
	}
	my $newimg = Imager->new();
	$newimg->read(%args) || die($errmsg . ": '" . $newimg->{ERRSTR} . "'");
	return $newimg;
}

# Sets the comparison method.  Either takes the numeric constant that
# identifies the method and any arguments needed by the method, or an instance
# of the comparator.  See the documentation for Image::Compare::Comparator or
# it subclasses for more details.
sub set_method {
	my $self = shift;
	my %args = @_;
	unless ($args{method}) {
		die "Missing required argument 'method'";
	}
	if (ref($args{method})) {
		if ($args{method}->isa('Image::Compare::Comparator')) {
			$self->{_CMP} = $args{method};
		}
		else {
			die (
				"Unrecognized type for 'method' argument: '" .
				ref($args{method}) . "'"
			);
		}
	}
	else {
		unless ($class_map{$args{method}}) {
			die "Unrecognized method identifier: '$args{method}'";
		}
		$self->{_CMP} = $class_map{$args{method}}->new($args{args});
	}
}

# Returns information describing the comparison method set into this instance
# of an Image::Compare.
sub get_method {
	my $self = shift;
	unless ($self->{_CMP}) {
		return wantarray ? () : undef;
	}
	return $self->{_CMP}->get_representation();
}

# Compares two images and returns a result.
sub compare {
	my $self;
	# This can be called as an instance method
	if (ref($_[0]) eq 'Image::Compare') {
		$self = shift;
	}
	else {
		# Or, as a class method, if you swing that way...
		if ($_[0] eq 'Image::Compare') {
			shift;
		}
		# Or just as a plain method.  In either case, we just need to construct
		# a $self so we can get on with life.
		$self = Image::Compare->new(@_);
	}
	# Sanity checking
	for my $ref (
		['IMG1', 'Image 1'], ['IMG2', 'Image 2'], ['CMP', 'Comparison method'],
	) {
		die "$ref->[1] not specified" unless $self->{"_$ref->[0]"};
	}

	# Give the images to the comparator and let them compare them.
	# The comparator will raise an exception if anything's wrong.
	return $self->{_CMP}->compare_images(
		@{$self}{qw/_IMG1 _IMG2 _MASK/}
	);
}

1;

__END__

=head1 NAME

Image::Compare - Compare two images in a variety of ways.

=head1 USAGE

 use Image::Compare;
 use warnings;
 use strict;

 my($cmp) = Image::Compare->new();
 $cmp->set_image1(
     img  => '/path/to/some/file.jpg',
     type => 'jpg',
 );
 $cmp->set_image2(
     img  => 'http://somesite.com/someimage.gif',
 );
 $cmp->set_method(
     method => &Image::Compare::THRESHOLD,
     args   => 25,
 );
 if ($cmp->compare()) {
     # The images are the same, within the threshold
 }
 else {
     # The images differ beyond the threshold
 }

=head1 OVERVIEW

This library implements a system by which 2 image files can be compared,
using a variety of comparison methods.  In general, those methods operate
on the images on a pixel-by-pixel basis and reporting statistics or data
based on color value comparisons.

C<Image::Compare> makes heavy use of the C<Imager> module, although it's not
neccessary to know anything about it in order to make use of the compare
functions.  However, C<Imager> must be installed in order to use this
module, and file import types will be limited to those supported by your
installed C<Imager> library.

In general, to do a comparison, you need to provide 3 pieces of information:
the first image to compare, the second image to compare, and a comparison
method.  Some comparison methods also require extra arguments -- in some cases
a boolean value, some a number and some require a hash reference with
structured data.  See the documentation below for information on how to use
each comparison method.

C<Image::Compare> provides 3 different ways to invoke its comparison
functionality -- you can construct an C<Image::Compare> object and call
C<set_*> methods on it to give it the information, then call C<compare()> on
that object, or you can construct the Image::Compare with all of the
appropriate data right off the bat, or you can simply call C<compare()>
with all of the information.  In this third case, you can call C<compare()>
as a class method, or you can simply invoke the method directly from the
C<Image::Compare> namespace.  If you'd like, you can also pass the word
C<compare> to the module when you C<use> it and the method will be
imported to your local namespace.

=head1 COMPARISON METHODS

=over 4

=item EXACT

The EXACT method simply returns true if every single pixel of one image
is exactly the same as every corresponding pixel in the other image, or false
otherwise.  It takes no arguments.

 $cmp->set_method(
     method => &Image::Compare::EXACT,
 );

=item THRESHOLD

The THRESHOLD method returns true if no pixel difference between the two images
exceeds a certain threshold, and false if even one does.  Note that differences
are measured in a sum of squares fashion (vector distance), so the maximum
difference is C<255 * sqrt(3)>, or roughly 441.7.  Its argument is the
difference threshold.  (Note:  EXACT is the same as THRESHOLD with an
argument of 0.)

 $cmp->set_method(
     method => &Image::Compare::THRESHOLD,
     args   => 50,
 );

=item THRESHOLD_COUNT

The THRESHOLD_COUNT method works similarly to the THRESHOLD method, but instead
of immediately returning a false value as soon as it finds a pixel pair whose
difference exceeds the threshold, it simply counts the number of pixels pairs
that exceed that threshold in the image pair. It returns that count.

 $cmp->set_method(
     method => &Image::Compare::THRESHOLD_COUNT,
     args   => 50,
 );

=item AVG_THRESHOLD

The AVG_THRESHOLD method returns true if the average difference over all pixel
pairings between the two images is under a given threshold value.  Two
different average types are available: MEDIAN and MEAN.  Its argument is a
hash reference, contains keys "type", indicating the average type, and
"value", indicating the threshold value.

 $cmp->set_method(
     method => &Image::Compare::AVG_THRESHOLD,
     args   => {
         type  => &Image::Compare::AVG_THRESHOLD::MEAN,
         value => 35,
     },
 );

=item IMAGE

The IMAGE method returns an C<Imager> object of the same dimensions as your
input images, with each pixel colored to represent the pixel color difference
between the corresponding pixels in the input.

Its only argument accepts 0, 1, or an L<Imager::Fountain>.  If the
argument is omitted or false, then the output image will be grayscale,
with black meaning no change and white meaning maximum change.  If the
argument is a true value other than an L<Imager::Fountain>, the output
will be in color, ramping from pure red at 0 change to pure green at
50% of maximum change, and then to pure blue at maximum change.

 $cmp->set_method(
     method => &Image::Compare::IMAGE,
     args   => 1,   # Output in color
 );

You may also pass an L<Imager::Fountain> to choose your own color scale.

 use Imager qw/:handy/; # for the NC subroutine
 use Imager::Fountain;

 $cmp->set_method(
     method => &Image::Compare::IMAGE,
     args   => Imager::Fountain->simple(
         positions => [              0.0,            1.0],
         colors    => [NC(255, 255, 255), NC(240,18,190)]
     )     # scale from white (no change) to fuschia (100% change)
 );

=back

=head1 MATCHMASKS

In addition to providing the two images which are to be compared, you may
also provide a "mask" image which will define a subset of those images to
compare.  A mask must be an Imager object, with one channel and 8 bit color
depth per channel.  Image processing will not occur for any pixel in the
test images which correspond to any pixel in the mask image with a color
value of (255, 255, 255), that is, black.

Put another way, the pure black section of the mask image effectively "hide"
that section of the test images, and those pixels will be ignored during
processing.  What that means will differ from comparator to comparator, but
should be obviously predictable in nature.

=head1 METHODS

=over 4

=item new()

=item new(image1 => { .. }, image2 => { .. }, method => { .. }, ..)

This is the constructor method for the class.  You may optionally pass it
any of 3 arguments, each of which takes a hash reference as data, which
corresponds exactly to the semantics of the C<set_*> methods, as described
below.  You may optionally pass in a match mask argument using the "mask"
argument, which must be an Imager object, as described above.

=item $cmp->set_image1(img => $data, type => $type)
=item $cmp->set_image2(img => $data, type => $type)

Sets the data for the appropriate image based on the input parameters.
The C<img> parameter can either be an C<Imager> object, a file path or a URL.
If a URL, it must be of a scheme supported by your C<LWP> install.  The C<type>
argument is optional, and will be used to override the image type deduced
from the input.  Again, the image type used must be one supported by your
C<Imager> install, and its format is determined entirely by C<Imager>.  See
the documentation on C<Imager::Files> for a list of image types.

Note that providing images as URLs requires that both LWP and Regexp::Common
be available in your kit.

=item $cmp->get_image1()
=item $cmp->get_image2()

Returns the underlying Imager object for the appropriate image, as created
inside of $cmp by either of the previous two methods.

=item $cmp->set_method(method => $method, args => $args)

Sets the comparison method for the object.  See the section above for details
on different comparison methods.

=item $cmp->get_method()

Returns a hash describing the method as set by the call previous.  In this
hash, the key "method" will map to the method, and the key "args" will map
to the arguments (if any).

=item $cmp->set_mask(mask => $mask)

Sets the match mask parameter as described above.

=item $cmp->get_mask()

Returns the match mask (if any) currently set in this object.

=item $cmp->compare()

=item compare(image1 => { .. }, image2 => { .. }, method => { .. })

Actually does the comparison.  The return value is determined by the comparison
method described in the previous section, so look there to see the details.
As described above, this can be called as an instance method, in which case
the values set at construction time or through the C<set_*> methods will be
used, or it can be called as a class method or as a simple subroutine.

In the latter case, all of the information must be provided as arguments to
the function call.  Those argument have exactly the same semantics as the
arguments for C<new()>, so see that section for details.

=back

=head1 Future Work

=over 4

=item *

I would like to implement more comparison methods.  I will have to use the
module myself somewhat before I know which ones would be useful to add, so
I'm releasing this initial version now with a limited set of comparisons.

I also more than welcome suggestions from users as to comparison methods
they would find useful, so please let me know if there's anything you'd like to
see the module be able to do.  This module is meant more to be a framework
for image comparison and a collection of systems working within that
framework, so the process of adding new comparison methods is reasonably
simple and painless.

=item *

I bet the input processing could be more bulletproof.  I am pretty certain of
it, in fact.

=item *

Maybe I could be more lenient with the format for masks.  I'll leave it up
to user request to see how I could extend that interface.

=back

=head1 Known Issues

=over 4

=item *

None at this time.

=back

=head1 AUTHOR

Copyright 2008 Avi Finkel <F<avi@finkel.org>>

This package is free software and is provided "as is" without express
or implied warranty.  It may be used, redistributed and/or modified
under the terms of the Perl Artistic License (see
http://www.perl.com/perl/misc/Artistic.html)

=cut
